File: libs/ktheme-swift/Sources/Ktheme/Presets/Presets.swift
// libs/ktheme-swift/Sources/Ktheme/Presets/Presets.swift
//
// Preset registry. The 24 canonical theme JSON files are bundled as
// resources in `Sources/Ktheme/Resources/themes/`. SwiftPM copies them
// into the module bundle.

import Foundation

public enum Presets {

    /// All 24 preset ids in the README's order.
    public static let allIds: [String] = [
        "navy-gold", "rose-gold", "burgundy-rose-gold", "emerald-silver",
        "midnight-amber", "deep-purple-platinum", "obsidian-crimson",
        "charcoal-champagne", "royal-silver", "royal-bronze",
        "slate-gunmetal", "slate-cyan", "forest-copper",
        "lcars", "windows-phone-metro", "frutiger-aero", "art-deco",
        "art-nouveau", "paper-ink",
        "solarpunk-civic", "calm-clinical", "neo-noir-neon",
        "aurora-glass-night", "ink-terminal-modern",
    ]

    private static var cache: [String: Theme] = [:]

    /// Load a single preset by id.
    public static func load(_ id: String) -> Theme {
        if let cached = cache[id] { return cached }
        guard let url = Bundle.module.url(
            forResource: id, withExtension: "json", subdirectory: "themes"
        ) else { preconditionFailure("Preset '\(id)' not bundled.") }
        do {
            let data = try Data(contentsOf: url)
            let theme = try ThemeParser.decode(data)
            cache[id] = theme
            return theme
        } catch {
            preconditionFailure("Failed to load preset '\(id)': \(error)")
        }
    }

    public static func all() -> [Theme] { allIds.map(load) }

    public static func byTag(_ tags: String...) -> [Theme] {
        let want = Set(tags)
        return all().filter { theme in !want.isDisjoint(with: Set(theme.metadata.tags)) }
    }
}

File: libs/ktheme-core/src/main/kotlin/io/ktheme/presets/Presets.kt
// libs/ktheme-core/src/main/kotlin/io/ktheme/presets/Presets.kt
//
// Presets registry. The 24 canonical theme JSON files ship inside the
// JAR under `themes/examples/*.json`. This object reads them lazily
// and exposes them by id.
//
// Build glue: the `ktheme-core` Gradle module is configured to copy
// the repo's `themes/examples/*.json` into `src/main/resources/
// themes/examples/` so they're packaged with the JAR. See
// `copyPresets` task in `build.gradle.kts`.

package io.ktheme.presets

import io.ktheme.model.Theme
import io.ktheme.parser.ThemeParser

public object Presets {

    /** All 24 preset ids in the order they appear in the README. */
    public val ALL_IDS: List<String> = listOf(
        // Metallic / elegant darks
        "navy-gold",
        "rose-gold",
        "burgundy-rose-gold",
        "emerald-silver",
        "midnight-amber",
        "deep-purple-platinum",
        "obsidian-crimson",
        "charcoal-champagne",
        "royal-silver",
        "royal-bronze",
        "slate-gunmetal",
        "slate-cyan",
        "forest-copper",
        // Iconic
        "lcars",
        "windows-phone-metro",
        "frutiger-aero",
        "art-deco",
        "art-nouveau",
        "paper-ink",
        // Product context
        "solarpunk-civic",
        "calm-clinical",
        "neo-noir-neon",
        "aurora-glass-night",
        "ink-terminal-modern",
    )

    private val cache = mutableMapOf<String, Theme>()

    /** Load a single preset by id. Reads from `themes/examples/{id}.json`. */
    public fun load(id: String): Theme = cache.getOrPut(id) {
        val path = "/themes/examples/$id.json"
        val stream = Presets::class.java.getResourceAsStream(path)
            ?: error("Preset '$id' not bundled. Expected resource at $path")
        ThemeParser.decode(stream.bufferedReader().use { it.readText() })
    }

    /** Load every preset listed in [ALL_IDS]. */
    public fun all(): List<Theme> = ALL_IDS.map(::load)

    /** Load every preset whose tags overlap [tags]. */
    public fun byTag(vararg tags: String): List<Theme> {
        val want = tags.toSet()
        return all().filter { theme -> theme.metadata.tags.any { it in want } }
    }
}
